In [16]:
import pandas as pd 
import numpy as np 
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings("ignore")

In [17]:
np.random.seed(1000)

# Entegable 2- Perceptrones y SVM.

Francisco Javier Mercader Martínez

## Machine Learning. Grado en Ciencia de datos

Los siguientes ejercicios tienen que ser entregados dentro de las dos horas de prácticas. Cómo realizar el entregable:


-   La realización se debe de hacer de forma  **INDIVIDUAL**
-   Se debe de enviar un notebook con el código y las explicaciones, comentarios, análisis y justificaciones en Markdown.
-   Reproducibilidad:
    -   Se debe de establecer una variable semilla con un número al inicio, esta variable será la que se utilice para el np.random_seed y para el random_state.
    -   Las bases de datos se cargan con rutas relativas.
-   El notebook se debe de subir a la tarea del aula virtual creada antes de la finalización de la hora de clase. (Ver tarea y fecha de cierre)

## Ejercicio 1

Utilizando el dataset Manzanas.csv, cree la SVM no lineal con un kernel de tipo función de base radial que considere para realizar la clasificación de la columna Calidad. 


Para ello, siga los siguientes pasos:

1. Cargue el conjunto de datos o dataset. Utiliza como objetivo o "y" la columna de *'Calidad'* y el resto de columnas como características.

In [18]:
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.svm import SVC
from sklearn import metrics

dataset = pd.read_csv("Manzanas.csv")

X = dataset.drop('Calidad', axis=1)
X = X.values

y = dataset['Calidad']

2. Cree un hold-out (ent-test) 80-20.

In [19]:
# Seleccionamos un 80% de los datos para entrenamiento y un 20% para test
X_ent, X_test, y_ent, y_test  = train_test_split(X, y, test_size=0.2)

3. Aplique la estandarización de los datos para ajustar la escala de las características.

In [20]:
# Instanciamos el escalador
scaler = StandardScaler()

# Estandarización del conjunto de entrenamiento
X_ent_escaladas = scaler.fit_transform(X_ent)

# Aplicamos la estandarización calculada con los datos de entrenamiento a los datos de test
X_test_escaladas = scaler.transform(X_test)

4. Cree una SVM no lineal con un kernel de tipo función de base radial ajustando tres hiperparámetros principales para conseguir una exactitud, precisión y sensibilidad por encima del 70% en el conjunto de entrenamiento y de test.

In [21]:
from sklearn.metrics import accuracy_score, precision_score, recall_score

# Instanciamos la clase LinearSVC
SVM_clasif_no_lineal = SVC(kernel='rbf', C=1.0, gamma='scale')

# Entrenamos la SVM kernelizada no lineal
SVM_clasif_no_lineal.fit(X_ent_escaladas, y_ent)

y_pred_ent = SVM_clasif_no_lineal.predict(X_ent_escaladas)
y_pred_test = SVM_clasif_no_lineal.predict(X_test_escaladas)

# Calcular y mostrar las métricas
print("Exactitud Entrenamiento: ", accuracy_score(y_ent, y_pred_ent))
print("Exactitud Test: ", accuracy_score(y_test, y_pred_test))
print("Precisión Entrenamiento: ", precision_score(y_ent, y_pred_ent, average='weighted'))
print("Precisión Test: ", precision_score(y_test, y_pred_test, average='weighted'))
print("Sensibilidad Entrenamiento: ", recall_score(y_ent, y_pred_ent, average='weighted'))
print("Sensibilidad Test: ", recall_score(y_test, y_pred_test, average='weighted'))

Exactitud Entrenamiento:  0.8959375
Exactitud Test:  0.885
Precisión Entrenamiento:  0.8959887650906927
Precisión Test:  0.8854647183800349
Sensibilidad Entrenamiento:  0.8959375
Sensibilidad Test:  0.885


5. ¿Qué influencia tiene el parámetro gamma ($\gamma$) en la capacidad de la SVM en el problema de clasificación? Para responder a esta pregunta realice varios cambios en el valor de gamma ($\gamma$) manteniendo los valores del resto de hiperparámetros y obtenga las métricas del ejercicio anterior en cada ejecución.

In [22]:
gammas = [0.1, 1, 10, 100]

for gamma in gammas:
    # Instanciamos la clase LinearSVC
    SVM_clasif_no_lineal = SVC(kernel='rbf', C=1.0, gamma=gamma)

    # Entrenamos la SVM kernelizada no lineal
    SVM_clasif_no_lineal.fit(X_ent_escaladas, y_ent)

    y_pred_ent = SVM_clasif_no_lineal.predict(X_ent_escaladas)
    y_pred_test = SVM_clasif_no_lineal.predict(X_test_escaladas)

    # Calcular y mostrar las métricas
    print(f"Gamma: {gamma}")
    print("Exactitud Entrenamiento: ", accuracy_score(y_ent, y_pred_ent))
    print("Exactitud Test: ", accuracy_score(y_test, y_pred_test))
    print("Precisión Entrenamiento: ", precision_score(y_ent, y_pred_ent, average='weighted'))
    print("Precisión Test: ", precision_score(y_test, y_pred_test, average='weighted'))
    print("Sensibilidad Entrenamiento: ", recall_score(y_ent, y_pred_ent, average='weighted'))
    print("Sensibilidad Test: ", recall_score(y_test, y_pred_test, average='weighted'))
    print("\n")

Gamma: 0.1
Exactitud Entrenamiento:  0.891875
Exactitud Test:  0.8875
Precisión Entrenamiento:  0.891932436308513
Precisión Test:  0.8879614280610935
Sensibilidad Entrenamiento:  0.891875
Sensibilidad Test:  0.8875


Gamma: 1
Exactitud Entrenamiento:  0.986875
Exactitud Test:  0.8875
Precisión Entrenamiento:  0.9868954051355797
Precisión Test:  0.8874658613941137
Sensibilidad Entrenamiento:  0.986875
Sensibilidad Test:  0.8875


Gamma: 10
Exactitud Entrenamiento:  1.0
Exactitud Test:  0.4725
Precisión Entrenamiento:  1.0
Precisión Test:  0.704905807120636
Sensibilidad Entrenamiento:  1.0
Sensibilidad Test:  0.4725


Gamma: 100
Exactitud Entrenamiento:  1.0
Exactitud Test:  0.46125
Precisión Entrenamiento:  1.0
Precisión Test:  0.2127515625
Sensibilidad Entrenamiento:  1.0
Sensibilidad Test:  0.46125




Se puede aprecia que conforme aumenta el valor de $\gamma$ los valores de las métricas del entrenamiento son cada vez mayores a diferencia que las métricas del test, cuyo valor va disminuyendo conforme aumenta el valor de $\gamma$ aunque sean diferencias muy pequeñas a nivel numérico.

6. Realice una operación de validación cruzada con 5 subconjuntos o folds con la SVM que ha generado en el ejercicio 4. Obtenga el acierto de cada subconjunto, la media del acierto y la desviación estándar del acierto.

In [23]:
from sklearn.model_selection import cross_val_score

puntuaciones = cross_val_score(SVM_clasif_no_lineal, X, y, cv=5)

print("Acierto de cada subconjunto: ", puntuaciones)
print("Media del acierto: ", np.mean(puntuaciones))
print("Desviación estándar del acierto: ", np.std(puntuaciones))

Acierto de cada subconjunto:  [0.50125 0.50125 0.50125 0.50125 0.5    ]
Media del acierto:  0.501
Desviación estándar del acierto:  0.0004999999999999894


7.  A la vista de los resultados, ¿es un clasificador robusto, fiable? Razone la respuesta.

Los resultados muestran que el acierto del clasificador es aproximadamente el mismo para cada subconjunto en la validación cruzada, con una media de 0.501 y una desviación estándar muy baja (<0.0005). Esto indica que el clasificador es bastante consistente en su rendimiento a través de diferentes subconjuntos de los datos.

Sin embargo, un acierto de alrededor del 50% no es particularmente alto. Por lo tanto, aunque este clasificador es consistente, no parece ser muy efectivo para este conjunto de datos en particular.

Nota: este dataset sirve para intentar predecir la calidad de las manzanas a partir de una serie de características: 'Size', 'Weight'	'Sweetness', 'Crunchiness',	'Juiciness', 'Ripeness','Acidity'. Utilizaremos todas las características para predecir la calidad.

## Ejercicio 2

Utilizando el dataset Mercado_valores_USA.csv, cree el perceptrón multicapa que considere para realizar la predicción de la columna *'S&P_500_Price'*. 


Para ello, siga los siguientes pasos:

1. Cargue el conjunto de datos o dataset. Seleccione como características: *'Natural_Gas_Price'*, *'Crude_oil_Price'*, *'Copper_Price'*. Seleccione como objetivo o "y" la columna *'S&P_500_Price'*.

In [30]:
mercado = pd.read_csv("Mercado_valores_USA.csv")

X = mercado[["Natural_Gas_Price", "Crude_oil_Price", "Copper_Price"]]
X = X.values

y = mercado["S&P_500_Price"].values

2. Cree un hold-out (ent-test) 80-20.

In [31]:
X_ent, X_test, y_ent, y_test = train_test_split(X, y, test_size=0.2)

3. Aplique la estandarización de los datos para ajustar la escala.

In [32]:
from sklearn.preprocessing import StandardScaler


scaler = StandardScaler()


X_ent_escaladas = scaler.fit_transform(X_ent)
X_test_escaladas = scaler.transform(X_test)

4. Cree un perceptrón multicapa de una única capa ajustando tres hiperparámetros principales para conseguir un error absoluto medio de entrenamiento y de test por debajo de 200. Obtenga también el error cuadrático medio. ¿Presenta algún problema de sobreajuste? Razone la respuesta.

In [33]:
from sklearn.neural_network import MLPRegressor
from sklearn.metrics import mean_absolute_error, mean_squared_error

perceptron_multicapa_reg = MLPRegressor(hidden_layer_sizes=(100,), max_iter=1000, random_state=42)
perceptron_multicapa_reg.fit(X_ent, y_ent)

y_pred_ent = perceptron_multicapa_reg.predict(X_ent_escaladas)
y_pred_test= perceptron_multicapa_reg.predict(X_test_escaladas)

print(f"Error absoluto medio de entrenamiento: {mean_absolute_error(y_ent, y_pred_ent)}")
print(f"Error cuadratico medio de entrenamiento: {mean_squared_error(y_ent, y_pred_ent)}")
print(f"Error absoluto medio de test: {mean_absolute_error(y_test, y_pred_test)}")
print(f"Error cuadratico medio de test: {mean_squared_error(y_test, y_pred_test)}")

Error absoluto medio de entrenamiento: 3576.2118933436395
Error cuadratico medio de entrenamiento: 13022954.210593564
Error absoluto medio de test: 3525.5893396685483
Error cuadratico medio de test: 12659747.804724341


5. Ejecute varias veces el entrenamiento del perceptrón multicapa utilizado, sin cambiar los valores de los hiperparámetros del perceptrón. En cada ejecución cambie la semilla. Obtenga las métricas pedidas en el ejercicio anterior. ¿Cambia el valor de las métricas? ¿Por qué? Razone la respuesta.

In [ ]:
semillas = [0, 1, 2, 3, 4]

for semilla in semillas:
    perceptron_multicapa_reg = MLPRegressor(hidden_layer_sizes=(100,), max_iter=1000, random_state=semilla)
    perceptron_multicapa_reg.fit(X_ent, y_ent)

    y_pred_ent = perceptron_multicapa_reg.predict(X_ent_escaladas)
    y_pred_test= perceptron_multicapa_reg.predict(X_test_escaladas)

    print(f"Semilla: {semilla}")
    print(f"Error absoluto medio de entrenamiento: {mean_absolute_error(y_ent, y_pred_ent)}")
    print(f"Error cuadratico medio de entrenamiento: {mean_squared_error(y_ent, y_pred_ent)}")
    print(f"Error absoluto medio de test: {mean_absolute_error(y_test, y_pred_test)}")
    print(f"Error cuadratico medio de test: {mean_squared_error(y_test, y_pred_test)}")
    print("\n")

Semilla: 0
Error absoluto medio de entrenamiento: 3538.490558949749
Error cuadratico medio de entrenamiento: 12736539.137807777
Error absoluto medio de test: 3490.8563674219286
Error cuadratico medio de test: 12397331.30349193


Semilla: 1
Error absoluto medio de entrenamiento: 3542.45511144773
Error cuadratico medio de entrenamiento: 12765505.081420258
Error absoluto medio de test: 3494.6057216685244
Error cuadratico medio de test: 12424439.856253702


Semilla: 2
Error absoluto medio de entrenamiento: 3555.209307871733
Error cuadratico medio de entrenamiento: 12862367.06645898
Error absoluto medio de test: 3506.422626218063
Error cuadratico medio de test: 12513709.503436774


Semilla: 3
Error absoluto medio de entrenamiento: 3563.623972000638
Error cuadratico medio de entrenamiento: 12926501.685195856
Error absoluto medio de test: 3514.1116082722633
Error cuadratico medio de test: 12572099.87208969


Semilla: 4
Error absoluto medio de entrenamiento: 3580.976425169099
Error cuadratico 

6. Realice una operación de búsqueda de los mejores parámetros. Emplee una y dos capas, utilice hasta dos números diferentes de neuronas en cada capa. Utilice dos funciones de activación. Utilice como método de cálculo de los pesos únicamente *'lbfgs'* y un valor máximo de iteraciones igual a 2000.

In [ ]:
from sklearn.model_selection import GridSearchCV
from sklearn.neural_network import MLPRegressor

# Definimos los parámetros a buscar
param_grid = {
    'hidden_layer_sizes': [(50,), (100,), (50, 50), (100, 100)], # Una y dos capas con diferentes números de neuronas
    'activation': ['tanh', 'relu'], # Funciones de activación
    'solver': ['lbfgs'], # Método de cálculo de los pesos
    'max_iter': [2000], # Número máximo de iteraciones
}

# Instanciamos el modelo
mlp = MLPRegressor(random_state=42)

# Configuramos la búsqueda de hiperparámetros
grid_search = GridSearchCV(mlp, param_grid, cv=5, scoring='neg_mean_absolute_error', verbose=1)

# Ejecutamos la búsqueda de hiperparámetros
grid_search.fit(X_ent_escaladas, y_ent)

# Mostramos los mejores parámetros
print(f"Mejores parámetros: {grid_search.best_params_}")

Fitting 5 folds for each of 8 candidates, totalling 40 fits


7. Obtenga la mejor configuración y su acierto. Dado el acierto obtenido, ¿obtiene la configuración obtenida un rendimiento mejor que la configuración generada en el ejercicio 4? Razone la respuesta. ¿Qué influencia tienen los parámetros que han cambiado en el resultado?

In [ ]:
print(f"Mejor configuración: {grid_search.best_estimator_}")
print(f"Mejor puntuación: {grid_search.best_score_}")

NameError: name 'grid_search' is not defined

Nota: este dataset sirve para intentar predecir el valor del índice bursátil S&P 500 a partir de una serie de características como: precio del gas natural, precio del crudo, precio del cobre, precio de ciertas accione, etc. Únicamente vamos a emplear las siguientes características: 'Natural_Gas_Price', 'Crude_oil_Price', 'Copper_Price'. 